demo/actions.py

```diff
@@ -17,35 +17,172 @@
 logger = logging.getLogger(__name__)
 
 
-class ActionSubscribeNewsletter(Action):
-    """ This action calls our newsletter API and subscribes the user with
-    their email address"""
+class SubscribeNewsletterForm(FormAction):
+    """Asks for the user's email, call the newsletter API and sign up user"""
 
     def name(self):
-        return "action_subscribe_newsletter"
+        return "subscribe_newsletter_form"
+
+    @staticmethod
+    def required_slots(tracker):
+        return ["email"]
+
+    def slot_mappings(self):
+        return {"email": self.from_entity(entity="email")}
+
+    def validate(self, dispatcher, tracker, domain):
+        # type: (CollectingDispatcher, Tracker, Dict[Text, Any]) -> List[Dict]
+        """Validate extracted value of requested slot
+            else reject execution of the form action
+
+            If no extraction, re-ask only if intent is enter_data
+        """
+        # extract other slots that were not requested
+        # but set by corresponding entity or trigger intent mapping
+        slot_values = self.extract_other_slots(dispatcher, tracker, domain)
+
+        # extract requested slot
+        slot_to_fill = tracker.get_slot(REQUESTED_SLOT)
+        if slot_to_fill:
+            slot_values.update(self.extract_requested_slot(dispatcher,
+                                                           tracker,
+                                                           domain))
+
+            if not slot_values:
+
+                # if no email entity was picked up, but intent was
+                # enter_data, ask again instead of leaving the form
+                intent = tracker.latest_message['intent'].get('name')
+                if intent == "enter_data":
+                    dispatcher.utter_template('utter_no_email', tracker)
+                    return []
+
+                # reject to execute the form action
+                # if some slot was requested but nothing was extracted
+                # it will allow other policies to predict another action
+                raise ActionExecutionRejection(self.name(),
+                                               "Failed to extract slot {0} "
+                                               "with action {1}"
+                                               "".format(slot_to_fill,
+                                                         self.name()))
+
+            for slot, value in slot_values.items():
+                validate_func = getattr(self, "validate_{}".format(slot),
+                                        lambda *x: value)
+                slot_values[slot] = validate_func(value, dispatcher, tracker,
+                                                  domain)
```

Comment:
do we need this? i don't think we have a validate function for this single slot

Reply:
the reason i implemented this was that the validate function for the slot only validates if it picks it up. if duckling picks up an email, i'm pretty sure that it's already valid. what this one does is if no entity is picked up, asks for the email again (but only if the intent is `enter_data`, since i still want it to exit if someone chitchats, denies, etc)

Reply:
but we could take the second chunk with the validate_{} functions out yeah

Reply:
oh yeah i was referring to this specific chunk, the `validate_func` stuff